src/library/scala/collection/mutable/BitSet.scala

```diff
@@ -171,6 +182,113 @@ class BitSet(protected[collection] final var elems: Array[Long])
   override def zip[B](that: IterableOnce[B])(implicit @implicitNotFound(collection.BitSet.zipOrdMsg) ev: Ordering[(Int, B)]): SortedSet[(Int, B)] =
     super.zip(that)
 
+  override def addAll(xs: IterableOnce[Int]): this.type = xs match {
+    case bs: collection.BitSet => this |= bs
+    case range: Range =>
+      if (range.nonEmpty) {
+        val start = range.min
+        if (start >= 0) {
+          val end = range.max
+          val endIdx = end >> LogWL
+          ensureCapacity(endIdx)
+
+          if (range.step == 1 || range.step == -1) {
+            val startIdx = start >> LogWL
+            val wordStart = startIdx * BitSetOps.WordLength
+            val wordMask = -1L << (start - wordStart)
+
+            if (endIdx > startIdx) {
+              elems(startIdx) |= wordMask
+              java.util.Arrays.fill(elems, startIdx + 1, endIdx, -1L)
+              elems(endIdx) |= -1L >>> (BitSetOps.WordLength - (end - endIdx * BitSetOps.WordLength) - 1)
+            } else {
+              elems(endIdx) |= (wordMask & (-1L >>> (BitSetOps.WordLength - (end - wordStart) - 1)))
+            }
+            this
+          } else {
+            super.addAll(range)
+          }
+        } else {
+          super.addAll(range)
+        }
+      } else {
+        this
+      }
+
+    case sorted: collection.SortedSet[Int] =>
+      // if `sorted` is using the regular Int ordering, ensure capacity for the largest
+      // element up front to avoid multiple resizing allocations
+      if (sorted.nonEmpty) {
+        val ord = sorted.ordering
+        if (ord eq Ordering.Int) {
+          ensureCapacity(sorted.lastKey >> LogWL)
+        } else if (ord eq Ordering.Int.reverse) {
+          ensureCapacity(sorted.firstKey >> LogWL)
+        }
+        val iter = sorted.iterator
+        while (iter.hasNext) {
+          addOne(iter.next())
+        }
+      }
+
+      this
+
+    case pq: PriorityQueue[Int] =>
+      // if `pq` is using the regular Int ordering, ensure capacity for the largest
+      // element up front to avoid multiple resizing allocations
+
+      if (pq.nonEmpty) {
+        val ord = pq.ord
+        if (ord eq Ordering.Int) {
+          ensureCapacity(pq.head >> LogWL)
+        }
+        val iter = pq.iterator
+        while (iter.hasNext) {
+          addOne(iter.next())
+        }
+      }
+
+      this
+
+
+    case other => super.addAll(other)
+  }
+
+  override def subsetOf(that: collection.Set[Int]): Boolean = that match {
+    case bs: collection.BitSet =>
+      val thisnwords = this.nwords
+      val bsnwords = bs.nwords
+
+      if (thisnwords <= bsnwords) {
```

Comment:
I think the logic could be simplified here as follows.  First, if `thiswords > bsnwords`, we check all the high ones and return `bsnwords` as the remainder to check; otherwise we do nothing and use `thiswords` as the number to check.  Second, check from 0 up to that number.  This way you won't have the code duplication of checking the low-index entries in two places.